### PR TITLE
ubuntu/focal-23.1.x:cherry pick netplan99-cannot-use-default.patch

### LINKED ISSUE #4133: Ubuntu Focal VM loses network configuration due to failed netplan generate
# Bug report
VM's based on Ubuntu 20.04, from before 20.04.4 have netplan.io package < 0.103, so they do not support the "new" to: default introduced in [netplan#216](https://github.com/canonical/netplan/pull/216).
When unattended-upgrades are enabled, cloud-init is automatically updated to release 23.1, which includes #4031.
As a result, the netplan generate step fails during boot, failing the cloud-init service and losing network configuration on a second reboot.

## Steps to reproduce the problem
* Provision a VM based on Ubuntu 20.04.3 or earlier, using cloud-init to configure the network interfaces
* Wait for unattended-upgrades to run, or upgrade cloud-init manually
* Reboot
* Observe cloud-init service has failed
* Reboot again
* Observe network configuration is no longer present

## Environment details
- Cloud-init version: 23.1
- Operating System Distribution: Ubuntu 20.04, from before 20.04.4
- Cloud provider, platform or installer type: Triton Datacenter (SmartOS datasource)

## cloud-init logs

```
2023-05-09 16:13:19,483 - stages.py[INFO]: Applying network configuration from ds bringup=False: {'version': 1, 'config': [{'mtu': 1500, 'type': 'physical', 'name': 'net0', 'mac_address': '90:b8:d0:33:00:e8', 'subnets': [{'type': 'sta
tic', 'address': '192.168.40.252/24', 'gateway': '192.168.40.254'}]}, {'type': 'nameserver', 'address': ['192.168.9.225'], 'search': []}]}
2023-05-09 16:13:19,487 - util.py[DEBUG]: Writing to /run/cloud-init/sem/apply_network_config.once - wb: [644] 24 bytes
2023-05-09 16:13:19,488 - __init__.py[DEBUG]: Selected renderer 'netplan' from priority list: ['netplan', 'eni', 'sysconfig']
2023-05-09 16:13:19,495 - subp.py[DEBUG]: Running command ['netplan', 'info'] with allowed return codes [0] (shell=False, capture=True)
2023-05-09 16:13:19,612 - util.py[DEBUG]: Attempting to load yaml from string of length 210 with allowed root types (<class 'dict'>,)
2023-05-09 16:13:19,618 - util.py[DEBUG]: Writing to /etc/netplan/50-cloud-init.yaml - wb: [644] 729 bytes
2023-05-09 16:13:19,619 - subp.py[DEBUG]: Running command ['netplan', 'generate'] with allowed return codes [0] (shell=False, capture=True)
2023-05-09 16:13:19,727 - util.py[WARNING]: failed stage init-local
2023-05-09 16:13:19,727 - util.py[DEBUG]: failed stage init-local
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/cloudinit/cmd/main.py", line 767, in status_wrapper
    ret = functor(name, args)
  File "/usr/lib/python3/dist-packages/cloudinit/cmd/main.py", line 433, in main_init
    init.apply_network_config(bring_up=bring_up_interfaces)
  File "/usr/lib/python3/dist-packages/cloudinit/stages.py", line 941, in apply_network_config
    return self.distro.apply_network_config(
  File "/usr/lib/python3/dist-packages/cloudinit/distros/__init__.py", line 278, in apply_network_config
    self._write_network_state(network_state, renderer)
  File "/usr/lib/python3/dist-packages/cloudinit/distros/debian.py", line 142, in _write_network_state
    return super()._write_network_state(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/cloudinit/distros/__init__.py", line 167, in _write_network_state
    renderer.render_network_state(network_state)
  File "/usr/lib/python3/dist-packages/cloudinit/net/netplan.py", line 298, in render_network_state
    self._netplan_generate(run=self._postcmds)
  File "/usr/lib/python3/dist-packages/cloudinit/net/netplan.py", line 305, in _netplan_generate
    subp.subp(self.NETPLAN_GENERATE, capture=True)
  File "/usr/lib/python3/dist-packages/cloudinit/subp.py", line 335, in subp
    raise ProcessExecutionError(
cloudinit.subp.ProcessExecutionError: Unexpected error while running command.
Command: ['netplan', 'generate']
Exit code: 1
Reason: -
Stdout:
Stderr:
        (generate:587): GLib-WARNING **: 18:13:19.715: GError set over the top of a previous GError or uninitialized memory.
        This indicates a bug in someone's code. You must ensure an error is NULL before it's set.
        The overwriting error message was: /etc/netplan/50-cloud-init.yaml:20:13: Error in network definition: unicast route must include both a 'to' and 'via' IP
                    -   to: default
                    ^
        /etc/netplan/50-cloud-init.yaml:20:21: Error in network definition: invalid IP family '-1'
                    -   to: default
                            ^
```

### PR DESCRIPTION
**Pull request by TheRealFalcon**
Fixes #4133

### CODE REVIEW

File: debian/patches/netplan99-cannot-use-default.patch [hunk 1]
```diff
@@ -0,0 +1,167 @@
+Description: Retain compatible route definition for netplan 0.99
+ focal contains netplan v0.99 which is not compatible
+ with Routes' definitions containing `to: default`. This patch
+ retains the previous way to define routes.
+Author: james.falcon@canonical.com
+Origin: backport
+Forwarded: not-needed
```
Comment (blackboxsw) [inline debian/patches/netplan99-cannot-use-default.patch:7]: Given this is tracked in a Github issue now, and we didn't represent that in d/changelog, we minimally want a link to the upstream issue here in dep3 header format
```suggestion
Bug: https://github.com/canonical/cloud-init/issues/4133 
Forwarded: not-needed
```

Comment (TheRealFalcon) [reply debian/patches/netplan99-cannot-use-default.patch:7]: Good idea. I just made sure we had tracking via these GitHub links, but this is better tracking.

File: debian/changelog [hunk 2]
```diff
@@ -1,3 +1,10 @@
+cloud-init (23.1.2-0ubuntu0~20.04.2) focal; urgency=medium
+
+  * d/patches/netplan99-cannot-use-default.patch:
+    - Retain routes' definitions compatible with netplan 0.99
```
Comment (blackboxsw) [inline debian/changelog:4]: Let's adopt the same references for upstream issues that UA uses in d/changelog
```suggestion
    - Retain routes' definitions compatible with netplan 0.99 (Fixes: #4133)
```

**Comment by TheRealFalcon** [comment]
@blackboxsw I'll reflect these changes to the other PR as well.

**Review by blackboxsw — APPROVED** [review]
Content and functionally is great here.
 I can confirm failure path when launching ubuntu focal images  and  seeding the following supplemental config on the filesystem to tell lxd to render network per reboot

```
cat /etc/cloud/cloud.cfg.d/50-net-per-boot-default-routes.cfg
updates:
  network:
    when: ['boot']
network:
   version: 1
   config:
                   - type: physical
                     name: interface0
                     mac_address: '00:16:3e:01:03:29'
                     subnets:
                       - type: static
                         address: 192.168.23.14/27
                         gateway: 192.168.23.1
                       - type: static
                         address: 11.0.0.11/24
                         gateway: 11.0.0.1
```
### After first reboot w/ , I see network break with 23.1.2-0ubuntu0~20.04.1 and invalid 'to: default' in 50-cloud-init.yaml
```
: Error in network definition: unicast route must include both a 'to' and 'via' IP
                    -   to: default
                    ^
        /etc/netplan/50-clo
```
```
me@dev-f:~# grep to: /etc/netplan/50-cloud-init.yaml 
            -   to: default
            -   to: default
```

After upgrade to this patched focal deb 23.1.2-0ubuntu0~20.04.2 no tracebacks and valid `to: 0.0.0.0/0` network
```
me@dev-f:~$ sudo dpkg -i /cloud-init_23.1.2-0ubuntu0~20.04.2_all.deb 
me@dev-f:~$ sudo reboot
me@dev-f:~$ grep to: /etc/netplan/50-cloud-init.yaml 
            -   to: 0.0.0.0/0
            -   to: 0.0.0.0/0
```

**Pull request merged**